### PR TITLE
feat: add sleep timer

### PR DESCRIPTION
**Pull request by SuperMeepBoy**
I tried to implement a sleep timer because it is a request, it exists on many similar software.

You may have things to say about it because I'm a beginner in front development. Maybe there are bugs I haven't anticipated, or maybe there is code which is dirty/can be written in a better way.

If you spot bugs/dirty code, tell me, I'll work on it again, improve whatever bug/dirty code you spot, it will help me learn and improve. :blush:

### CODE REVIEW

**Comment by remvze** [comment]
This is great. I'll review it right now.

File: src/components/menu/menu.tsx [hunk 1]
```diff
@@ -64,13 +67,15 @@ export function Menu() {
...
   useHotkeys('shift+s', () => open('shareLink'), { enabled: !noSelected });
+  useHotkeys('shift+t', () => open('sleepTimer'));
 
   useCloseListener(closeAll);
 
   const variants = mix(fade(), slideY());
 
   return (
     <>
+      <Divider />
```
Comment (remvze) [inline src/components/menu/menu.tsx:78]: Is this left by accident? If not, what purpose does it serve? It creates a line between the two toolbar buttons that isn't aesthetically pleasing.

File: src/components/modals/sleep-timer/sleep-timer.tsx [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+            clearInterval(newTimerId);
+            pause();
+            setRunning(false);
+            return 0;
+          }
+          return newTimeLeft;
+        });
+      }, 1000);
+
+      setTimerId(newTimerId);
```
Comment (remvze) [inline src/components/modals/sleep-timer/sleep-timer.tsx:57]: It's better to use a ref instead of state for timerId.

**Review by remvze — CHANGES_REQUESTED** [review]

File: src/components/menu/menu.tsx (hunk 1 above)
Comment (SuperMeepBoy) [reply src/components/menu/menu.tsx:78]: Yup, accidental ! Sorry and thanks for noticing !

File: src/components/modals/sleep-timer/sleep-timer.tsx (hunk 2 above)
Comment (SuperMeepBoy) [reply src/components/modals/sleep-timer/sleep-timer.tsx:57]: Going to check that ASAP. I will put everything in a second commit so you don't have to review everything a second time. I let you squash them if you want !

**Comment by SuperMeepBoy** [comment]
@remvze Done it ! I think this is what you asked for. Tell me if there is something else I can do !

Also I wonder why ref was better that state. I tend to always use state but again, I have little experience so maybe I just don't really understand the difference and their use cases. If you have a good reference or if you can explain, it would be lovely ! :blush:

**Comment by SuperMeepBoy** [comment]
@remvze I fixed a few more things :

- The timer had a bug. I just reset the values to 0 when we click "reset". Maybe we could remember the previous hours/minutes set by the user. I don't really know ! But now, there is not the bug anymore
- We could start the timer with 00:00:00. I prevented that by disabling the start button
- We can start the timer even if the sounds aren't playing. I wanted to start sound if we do so, but then the user can just pause the sounds and the timer will still keep going so I don't really see the point. Maybe if someone asks for this, we can implement it later it is not a big deal but I'd rather keep it simple I guess :thinking: 
The option of disabling opening this modal if sounds aren't playing is not an option as again a user can start the timer, then manually pause the sounds and the timer would still be going on.

The feature seems ready to me and is as simple as it can be : just a timer which will pause the sound if they are playing. If they are not… nothing will happen.

**Review by remvze — APPROVED** [review]

**Comment by remvze** [comment]
Thank you so much. I'll merge this into the develop branch first, make a few adjustments that are harder to communicate here, then I'll merge it into the main branch. I really appreciate your contributions to this project.

**Pull request merged**